### PR TITLE
Feature/farm 279 ui update crop progress indicator

### PR DESCRIPTION
**Pull request by isaacmartinezimv**
[[FARM-279]-UI Update Crop Progress Indicator](https://amidodevelopment.atlassian.net/secure/RapidBoard.jspa?rapidView=150&projectKey=FARM&view=planning&selectedIssue=FARM-279)

#### 📲 What

Crop progress indicator in MyPlot page

#### 🤔 Why
		
To follow more accurately the progress of the crop stages.

#### 🛠 How
		
Using a custom painter to draw the circle and populate it with a function to be reusable.

#### 👀 See
		
CircularGif
		 
#### ✅ Acceptance criteria

- [X] Design Review for UI with BA completed. 
- [X] Manually tested and verified on Android device/emulator (min. Lollipop 5.1)
- [ ] Showcase video / automation test
- [X] Passing all tests
- [ ] Rebased/merged with latest changes from development and re-tested
- [ ] Removed unsed comments. TODOs must have JIRA for future work.

#### Coding Standards Checklist
- [ ] unit tests 80% coverage on testable code (functions, methods, class)
- [X] (Architecture) redux state, reducers, actions, middleware defined under lib/redux
- [ ] ui elements defined under lib/ui. 
- [ ] Strings ready for localisation (e.g. defined in lib/utils/strings.dart)
- [ ] Assets (images/icons path) defined in lib/utils/assets.dart
- [ ] Colors defined in lib/utils/colors.dart
- [ ] Dimensions (ie margins padding) defined in lib/utils/dimens.dart
- [ ] TextStyles defined in lib/utils/styles.dart

Recommended Style Guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

#### 🕵️‍♂️ How to test

Notes for QA

#### Reviewers

### CODE REVIEW

File: lib/ui/common/CircularProgress.dart [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        height: widget._style.height,
+        width: widget._style.width,
+        child: CustomPaint(
+          foregroundPainter: MyPainter(
+              lineColor: widget._style.lineColor,
+              completeColor: widget._style.completeColor,
+              completePercent: widget._viewModel.initialValue,
+              width: widget._style.lineWidth),
+          child: Padding(
+            padding: widget._style.innerPadding,
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: we talked about this not having children as an input and letting the point of use sort that out, was that not possible?

File: lib/ui/common/MyPainter.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
+
+
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+
+class MyPainter extends CustomPainter{
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/MyPainter.dart]: very poor name. What does it do?

**Review by LeeMatthewHiggins — CHANGES_REQUESTED** [review]
Lets have a pair programming session to go over this

File: lib/ui/common/MyPainter.dart (hunk 2 above)
Comment (isaacmartinezimv) [reply lib/ui/common/MyPainter.dart]: File deleted, now it's a private class in the CircularWidget file like we talked

File: lib/ui/common/CircularProgress.dart (hunk 1 above)
Comment (isaacmartinezimv) [reply lib/ui/common/CircularProgress.dart]: Now the widget is the simplest version, and it's filled in the use case situation with a Stack widget

File: lib/ui/common/MyPainter.dart (hunk 2 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/MyPainter.dart]: 👍

File: lib/ui/myplot/PlotListItem.dart [hunk 3]
```diff
@@ -119,9 +119,14 @@ class PlotListItem {
...
                         SizedBox(width: itemStyle.imageLineSpace),
-                        //FIXME: This value parameters are hardcoded right now, later should be the current stage number divided by all the stages.
-                        CircularProgress(viewModel: CircularProgressViewModel(
-                            initialValue: 20, content: _buildPlotImageContent(itemStyle, viewModel.imageUrl)))
+                        Stack(
+                          alignment: AlignmentDirectional.center,
+                          children: <Widget>[
+                            _buildPlotImage(viewModel.imageUrl, itemStyle),
+                            //FIXME: This value parameters are hardcoded right now, later should be the current stage number divided by all the stages.
+                            CircularProgress(progress: 80),
```
Comment (LeeMatthewHiggins) [inline lib/ui/myplot/PlotListItem.dart]: grab the value from the viewModel named  "progress"  double 0.0 - 1.0

File: lib/ui/playground/data/playground_atom_datasource.dart [hunk 4]
```diff
@@ -151,7 +151,7 @@ class PlayGroundAtomDataSource implements PlaygroundDataSource {
         child: Container(
           alignment: Alignment.center,
           padding: EdgeInsets.symmetric(vertical: 20.0),
-          child: CircularProgress(viewModel: MockCircularProgressViewModel.buildWithImage()),
+          child: CircularProgress(progress: 75)
```
Comment (LeeMatthewHiggins) [inline lib/ui/playground/data/playground_atom_datasource.dart]: use normalised terms 0.0 - 1.0

File: lib/ui/common/CircularProgress.dart [hunk 5]
```diff
@@ -102,22 +33,67 @@ class _CircularProgressState extends State<CircularProgress> {
...
+
+    Offset center = Offset(size.width/2, size.height/2);
+    double radius = min(size.width/2, size.height/2);
+    canvas.drawCircle(
+        center,
+        radius,
+        line
+    );
+
+    double arcAngle = 2*pi* (completePercent/100);
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: take in a normalised number, not per cent. that removes this divide. clamp the input to 0.0-1.0 so it will never bet outside this range here.

File: lib/ui/common/CircularProgress.dart [hunk 6]
```diff
@@ -91,6 +17,11 @@ class CircularProgress extends StatefulWidget {
 class _CircularProgressState extends State<CircularProgress> {
   double defaultValue = 0;
   double percentageComplete = 100.0;
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: try and use normalised terms (0.0-1.0). percent is a display implementation detail.

**Review by LeeMatthewHiggins — CHANGES_REQUESTED** [review]
minor changes and we are done 👍

File: lib/ui/myplot/PlotListItem.dart (hunk 3 above)
Comment (isaacmartinezimv) [reply lib/ui/myplot/PlotListItem.dart]: Fixed

File: lib/ui/common/CircularProgress.dart [hunk 7]
```diff
@@ -0,0 +1,99 @@
+import 'package:flutter/material.dart';
+import 'dart:math';
+
+class CircularProgress extends StatefulWidget {
+  final Color _lineColor;
+  final double _progress;
+
+  const CircularProgress({Key key, @required double progress, Color lineColor})
+      : this._lineColor = lineColor ?? const Color(0xff24d900),
+        this._progress = progress,
```
Comment (davidfranquet) [inline lib/ui/common/CircularProgress.dart:14]: Add an assert here :D
```suggestion
        this._progress = progress,
        assert(progress >= 0 && progress <= 1, 'Progress should be between 0 and 1'),
```

**Review by davidfranquet — DISMISSED** [review]

File: lib/ui/common/CircularProgress.dart [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        assert(progress >= 0 && progress <= 1, 'Progress should be between 0 and 1'),
+        super(key: key);
+
+  @override
+  _CircularProgressState createState() => _CircularProgressState();
+}
+
+class _CircularProgressState extends State<CircularProgress> {
+  double defaultValue = 0;
+  double percentageComplete = 1.0;
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: is it a percent?    maybe name "progress" now

File: lib/ui/common/CircularProgress.dart [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        this._progress = progress,
+        assert(progress >= 0 && progress <= 1, 'Progress should be between 0 and 1'),
+        super(key: key);
+
+  @override
+  _CircularProgressState createState() => _CircularProgressState();
+}
+
+class _CircularProgressState extends State<CircularProgress> {
+  double defaultValue = 0;
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: that used?

File: lib/ui/common/DogTagStyles.dart [hunk 10]
```diff
@@ -21,7 +21,7 @@ class DogTagStyles {
             EdgeInsets.only(left: 12, top: 5.5, right: 12, bottom: 5.5),
         titleTextStyle: TextStyle(
             fontSize: 11,
-            fontWeight: FontWeight.w500,
+            fontWeight: FontWeight.w600,
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/DogTagStyles.dart]: is this part of this PR?

File: lib/ui/myplot/PlotListItem.dart [hunk 11]
```diff
@@ -3,27 +3,29 @@ import 'package:farmsmart_flutter/ui/common/DogTagStyles.dart';
...
-      this.title, this.subtitle, this.detail, this.imageUrl, this.onTap);
+      this.title, this.subtitle, this.detail, this.imageUrl, this.onTap, this.progress);
 }
 
 PlotListItemViewModel fromCropEntityToViewModel(
     CropEntity currentCrop, Function goToDetail) {
   //FIXME: Change the mocked data "planting" and "Day 6" with the correct FirebaseData when available
   return PlotListItemViewModel(currentCrop.name ?? Strings.defaultCropNameText,
-      "Planting", "Day 6", currentCrop.imageUrl, () => goToDetail(currentCrop));
+      "Planting", "Day 6", currentCrop.imageUrl, () => goToDetail(currentCrop), 0.7);
```
Comment (LeeMatthewHiggins) [inline lib/ui/myplot/PlotListItem.dart]: 👍

**Review by LeeMatthewHiggins — CHANGES_REQUESTED** [review]

File: lib/ui/common/DogTagStyles.dart (hunk 10 above)
Comment (isaacmartinezimv) [reply lib/ui/common/DogTagStyles.dart]: Totally agree, just saw that the designs are a little bit more bold in that dogTag and i dont realised that is out of scope of this PR, my bad

**Review by davidfranquet — DISMISSED** [review]
Rebase with develop and solve the conflict 😜

File: lib/ui/common/CircularProgress.dart [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  final double height = 87;
+  final double width = 87;
+  final Color lineColor = Colors.transparent;
+  final double lineWidth = 3;
+  final EdgeInsets innerPadding = const EdgeInsets.all(4.0);
+
+  @override
+  void initState() {
+    super.initState();
+    setState(() {});
```
Comment (davidfranquet) [inline lib/ui/common/CircularProgress.dart]: Needed? 🤔

File: lib/ui/mockData/MockDogTagViewModel.dart [hunk 13]
```diff
@@ -1,5 +1,6 @@
 import 'package:farmsmart_flutter/ui/common/Dogtag.dart';
 import 'package:farmsmart_flutter/ui/common/MockString.dart';
+import 'package:flutter/material.dart';
```
Comment (davidfranquet) [inline lib/ui/mockData/MockDogTagViewModel.dart]: This is not part of this PR

**Review by davidfranquet — CHANGES_REQUESTED** [review]

File: lib/ui/common/CircularProgress.dart (hunk 12 above)
Comment (isaacmartinezimv) [reply lib/ui/common/CircularProgress.dart]: it will be needed

File: lib/ui/mockData/MockDogTagViewModel.dart (hunk 13 above)
Comment (davidfranquet) [reply lib/ui/mockData/MockDogTagViewModel.dart]: PRAssistant

**Review by davidfranquet — DISMISSED** [review]

File: lib/ui/common/CircularProgress.dart (hunk 12 above)
Comment (LeeMatthewHiggins) [reply lib/ui/common/CircularProgress.dart]: for what?

File: lib/ui/common/CircularProgress.dart [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+          child: Padding(
+            padding: innerPadding,
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class _CircularPainter extends CustomPainter {
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart:42]: perfect, implementation details hidden. 👍

File: lib/ui/common/CircularProgress.dart [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+  Color lineColor;
+  Color completeColor;
+  double completePercent;
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: is it a per cent? 0-100?   I think its better to be 0-1. else you will constantly be * by 100 in one place just to / by 100 in another.

File: lib/ui/common/CircularProgress.dart [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class _CircularPainter extends CustomPainter {
+  Color lineColor;
+  Color completeColor;
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/CircularProgress.dart]: I think lineColor is better name here, and trackColor above.  This makes me think it changes colour when complete.

**Review by LeeMatthewHiggins — DISMISSED** [review]
Two small naming things, once changed you can merge

**Review by davidfranquet — DISMISSED** [review]

**Review by LeeMatthewHiggins — APPROVED** [review]

**Review by davidfranquet — APPROVED** [review]
